### PR TITLE
update loadbalancers service to use Azure SDK v2

### LINKED ISSUE #3917: Update loadbalancers service to SDK v2
part of https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3409

### PR DESCRIPTION
**Pull request by nawazkh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- This PR updates load balancers service to sdkv2. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3917 

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Update loadbalancers service to use Azure SDK v2
```

### CODE REVIEW

**Comment by nawazkh** [comment]
I will squash it once the e2e go green :)

File: azure/converters/loadbalancers.go [hunk 1]
```diff
@@ -17,14 +17,14 @@ limitations under the License.
...
 
 import (
-	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-08-01/network"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork/v4"
 	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
 )
 
-// SKUtoSDK converts infrav1.SKU into a network.LoadBalancerSkuName.
-func SKUtoSDK(src infrav1.SKU) network.LoadBalancerSkuName {
+// SKUtoSDK converts infrav1.SKU into a armnetwork.LoadBalancerSkuName.
```
Comment (mboersma) [inline azure/converters/loadbalancers.go:24]:
```suggestion
// SKUtoSDK converts infrav1.SKU into an armnetwork.LoadBalancerSkuName.
```

File: azure/services/loadbalancers/client.go [hunk 2]
```diff
@@ -18,158 +18,127 @@ package loadbalancers
...
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "loadbalancers.azureClient.Get")
 	defer done()
 
-	return ac.loadbalancers.Get(ctx, spec.ResourceGroupName(), spec.ResourceName(), "")
+	resp, err := ac.loadbalancers.Get(ctx, spec.ResourceGroupName(), spec.ResourceName(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp, nil
```
Comment (mboersma) [inline azure/services/loadbalancers/client.go:62]:
```suggestion
	return resp.LoadBalancer, nil
```

This is what needs to be returned, but unfortunately the signature is just `interface{}` so we can't enforce the return value.

File: azure/services/loadbalancers/client.go [hunk 3]
```diff
@@ -18,158 +18,127 @@ package loadbalancers
...
-		err = autorest.NewErrorWithError(err, "network.LoadBalancersClient", "CreateOrUpdate", nil, "Failure preparing request")
-		return nil, nil, err
+		return nil, nil, errors.Wrap(err, "failed to create loadbalancer client options")
 	}
 
-	if etag != "" {
-		req.Header.Add("If-Match", etag)
+	factory, err := armnetwork.NewClientFactory(ac.auth.SubscriptionID(), ac.auth.Token(), clientOpts)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to create loadbalancer client factory")
```
Comment (mboersma) [inline azure/services/loadbalancers/client.go:94]:
```suggestion
		return nil, nil, errors.Wrap(err, "failed to create amrnetwork client factory")
```

**Review by mboersma — CHANGES_REQUESTED** [review]

File: azure/converters/loadbalancers.go (hunk 1 above)
Comment (nawazkh) [reply azure/converters/loadbalancers.go]: Updated

File: azure/services/loadbalancers/client.go (hunk 2 above)
Comment (nawazkh) [reply azure/services/loadbalancers/client.go]: Updated

File: azure/services/loadbalancers/client.go (hunk 3 above)
Comment (nawazkh) [reply azure/services/loadbalancers/client.go]: Updated

**Comment by nawazkh** [comment]
Expected complaints from apidiff
```
sigs.k8s.io/cluster-api-provider-azure/azure/services/loadbalancers
Incompatible changes:
  - New: changed from func(LBScope) *Service to func(LBScope) (*Service, error)

sigs.k8s.io/cluster-api-provider-azure/azure/converters
Incompatible changes:
  - SKUtoSDK: changed from func(sigs.k8s.io/cluster-api-provider-azure/api/v1beta1.SKU) github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-08-01/network.LoadBalancerSkuName to func(sigs.k8s.io/cluster-api-provider-azure/api/v1beta1.SKU) github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork/v4.LoadBalancerSKUName
```

File: azure/converters/loadbalancers.go [hunk 4]
```diff
@@ -17,14 +17,14 @@ limitations under the License.
...
 
 import (
-	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-08-01/network"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork/v4"
 	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
 )
 
-// SKUtoSDK converts infrav1.SKU into a network.LoadBalancerSkuName.
-func SKUtoSDK(src infrav1.SKU) network.LoadBalancerSkuName {
+// SKUtoSDK converts infrav1.SKU into an armnetwork.LoadBalancerSkuName.
```
Comment (nojnhuh) [inline azure/converters/loadbalancers.go:24]: nit:
```suggestion
// SKUtoSDK converts infrav1.SKU into an armnetwork.LoadBalancerSKUName.
```

File: azure/services/loadbalancers/client.go [hunk 5]
```diff
@@ -18,158 +18,127 @@ package loadbalancers
...
 )
 
 // azureClient contains the Azure go-sdk Client.
 type azureClient struct {
-	loadbalancers network.LoadBalancersClient
+	loadbalancers *armnetwork.LoadBalancersClient
+	auth          azure.Authorizer
 }
 
 // newClient creates a new load balancer client from subscription ID.
```
Comment (nojnhuh) [inline azure/services/loadbalancers/client.go:38]: nit:
```suggestion
// newClient creates a new load balancer client from an authorizer.
```

Comment (nawazkh) [reply azure/services/loadbalancers/client.go]: Updated

File: azure/converters/loadbalancers.go (hunk 4 above)
Comment (nawazkh) [reply azure/converters/loadbalancers.go]: Updated.

File: azure/services/loadbalancers/client.go [hunk 6]
```diff
@@ -18,158 +18,127 @@ package loadbalancers
...
+	}
 
-// newLoadbalancersClient creates a new load balancer client from subscription ID.
-func newLoadBalancersClient(subscriptionID string, baseURI string, authorizer autorest.Authorizer) network.LoadBalancersClient {
-	loadBalancersClient := network.NewLoadBalancersClientWithBaseURI(baseURI, subscriptionID)
-	azure.SetAutoRestClientDefaults(&loadBalancersClient.Client, authorizer)
-	return loadBalancersClient
+	factory, err := armnetwork.NewClientFactory(auth.SubscriptionID(), auth.Token(), opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create load balancer client factory")
```
Comment (mboersma) [inline azure/services/loadbalancers/client.go:47]:
```suggestion
		return nil, errors.Wrap(err, "failed to create armnetwork client factory")
```

**Review by mboersma — COMMENTED** [review]
lgtm, I just had one change to a comment for consistency with the other services.

File: azure/services/loadbalancers/client.go (hunk 6 above)
Comment (nawazkh) [reply azure/services/loadbalancers/client.go]: Updated :)

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

**Pull request merged**